### PR TITLE
Make AffiliateValidator a trusted contract

### PR DESCRIPTION
**Pull request by nuevoalex**

### CODE REVIEW

File: packages/augur-core/source/contracts/reporting/Affiliates.sol [hunk 1]
```diff
@@ -10,9 +11,15 @@ contract Affiliates {
     // Maps an account to the referral account. Used to apply affiliate fees on settlement.
     mapping (address => address) public referrals;
 
-    uint256 constant internal AFFILIATE_VALIDATOR_GAS = 30000;
+    // Mapping of valid Affiliate Validators
+    mapping (address => bool) public affiliateValidators;
 
-    bytes4 constant internal AFFILIATE_VALIDATOR_SELECTOR = bytes4(keccak256("validateReference(address,address)"));
+    function createAffiliateValidator() public returns (AffiliateValidator) {
+        AffiliateValidator _affiliateValidator = new AffiliateValidator();
```
Comment (epheph) [inline packages/augur-core/source/contracts/reporting/Affiliates.sol:18]: This whole thing feels way better. I doubt a user would need to implement custom logic on-chain for affiliates.

File: packages/augur-core/source/contracts/trading/ZeroXTrade.sol [hunk 2]
```diff
@@ -159,14 +159,16 @@ contract ZeroXTrade is Initializable, IZeroXTrade, IERC1155 {
...
+        uint256 _availableCash = cash.allowance(_owner, augur).min(cash.balanceOf(_owner));
+        uint256 _attoSharesPurchasable = _availableCash.div(_price);
 
         return _attoSharesOwned.add(_attoSharesPurchasable);
     }
 
     function askBalance(address _owner, IMarket _market, uint8 _outcome, uint256 _price) public view returns (uint256) {
         uint256 _attoSharesOwned = shareToken.balanceOfMarketOutcome(_market, _outcome, _owner);
-        uint256 _attoSharesPurchasable = cash.balanceOf(_owner).div(_market.getNumTicks().sub(_price));
+        uint256 _availableCash = cash.allowance(_owner, augur).min(cash.balanceOf(_owner));
```
Comment (epheph) [inline packages/augur-core/source/contracts/trading/ZeroXTrade.sol]: This should be the allowance of FillOrder and not augur

Comment (nuevoalex) [reply packages/augur-core/source/contracts/trading/ZeroXTrade.sol]: Yup! Thanks for the catch

File: packages/augur-core/source/contracts/reporting/AffiliateValidator.sol [hunk 3]
```diff
@@ -45,13 +45,13 @@ contract AffiliateValidator is Ownable, IAffiliateValidator {
...
-    function validateReference(address _account, address _referrer) external view {
+    function validateReference(address _account, address _referrer) external view returns (bool) {
         bytes32 _accountKey = keys[_account];
         bytes32 _referralKey = keys[_referrer];
         if (_accountKey == bytes32(0) || _referralKey == bytes32(0)) {
-            revert("Key must be registered for both accounts");
+            return false;
         }
-        require(_accountKey != _referralKey, "Key must not be equal for both accounts");
+        return _accountKey != _referralKey;
```
Comment (epheph) [inline packages/augur-core/source/contracts/reporting/AffiliateValidator.sol:54]: I'm glad you changed this, there was a troll vector here before

File: packages/augur-core/source/contracts/trading/ZeroXTrade.sol (hunk 2 above)
Comment (epheph) [reply packages/augur-core/source/contracts/trading/ZeroXTrade.sol]: I wish there was a standard for "how much can i transferFrom" without calling into the contract twice.

**Review by epheph — APPROVED** [review]
This feels a lot safer. There is a change in here not related affiliate, the fill order balanceOf lookups that that prevent revert's when balance exceeds FillOrder's allowance.

**Pull request merged**